src/Analyzers/Core/Analyzers/ForEachCast/AbstractForEachCastDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,148 @@
+        where TSyntaxKind : struct, Enum
+        where TForEachStatementSyntax : SyntaxNode
+    {
+        public static readonly ImmutableDictionary<string, string?> s_isFixableProperties =
+            ImmutableDictionary<string, string?>.Empty.Add(ForEachCastHelpers.IsFixable, ForEachCastHelpers.IsFixable);
+
+        protected AbstractForEachCastDiagnosticAnalyzer(string language)
+            : base(
+                  diagnosticId: IDEDiagnosticIds.ForEachCastDiagnosticId,
+                  EnforceOnBuildValues.ForEachCast,
+                  option: null,
+                  language: language,
+                  title: new LocalizableResourceString(nameof(AnalyzersResources.Add_explicit_cast), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
+                  messageFormat: new LocalizableResourceString(nameof(AnalyzersResources._0_statement_implicitly_converts_1_to_2_and_may_fail_at_runtime), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        protected abstract ImmutableArray<TSyntaxKind> GetSyntaxKinds();
+        protected abstract (CommonConversion conversion, ITypeSymbol? collectionElementType) GetForEachInfo(SemanticModel semanticModel, TForEachStatementSyntax node);
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var ienumerableType = compilation.IEnumerableType();
+                var ienumerableOfTType = compilation.IEnumerableOfTType();
+                if (ienumerableType != null && ienumerableOfTType != null)
+                    context.RegisterSyntaxNodeAction(context => AnalyzeSyntax(context, ienumerableType, ienumerableOfTType), GetSyntaxKinds());
+            });
+        }
+
+        protected void AnalyzeSyntax(
+            SyntaxNodeAnalysisContext context, INamedTypeSymbol ienumerableType, INamedTypeSymbol ienumerableOfTType)
+        {
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+            if (context.Node is not TForEachStatementSyntax node)
+                return;
+
+            var option = context.GetOption(CodeStyleOptions2.ForEachExplicitCastInSource, semanticModel.Language);
+            if (option.Value == ForEachExplicitCastInSourcePreference.Never)
+                return;
+
+            Contract.ThrowIfFalse(option.Value is ForEachExplicitCastInSourcePreference.Always or ForEachExplicitCastInSourcePreference.NonLegacy);
+
+            if (semanticModel.GetOperation(node, cancellationToken) is not IForEachLoopOperation loopOperation)
```

Comment:
Why we're not registering an operation action from the beginning?

Reply:
On Discord:

![image](https://user-images.githubusercontent.com/31348972/157969537-66cd5d1f-71dd-4e64-8961-0da22d094eb9.png)

> i prefer syntax as i know exactly what i'm getting. we coudl go either direction.

Both operation and syntax checks already exist. It's a matter of what we registered callbacks for. @sharwell Does it matter for this analyzer what we register callbacks for?

Reply:
i mean, it really depends.  i agree with sam that for parameter<->argument mapping IOp is the way to go.  Here it's really not relevant.  We can go from one to the other trivially.